src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+            const type = range.facts & RangeFacts.InStaticRegion ? checker.getTypeOfSymbolAtLocation(scope.symbol, scope) : checker.getDeclaredTypeOfSymbol(scope.symbol);
+            const props = checker.getPropertiesOfType(type);
+            functionNameText = getUniqueName(n => props.every(p => p.name !== n));
+        }
+        else {
+            const file = scope.getSourceFile();
+            functionNameText = getUniqueName(n => !file.identifiers.has(n as string));
+        }
+        const isJS = isInJavaScriptFile(scope);
+
+        const functionName = createIdentifier(functionNameText as string);
+        const functionReference = createIdentifier(functionNameText as string);
+
+        let returnType: TypeNode = undefined;
+        const parameters: ParameterDeclaration[] = [];
+        const callArguments: Identifier[] = [];
+        let writes: UsageEntry[];
+        usagesInScope.forEach((usage, name) => {
+            let typeNode: TypeNode = undefined;
+            if (!isJS) {
+                let type = checker.getTypeOfSymbolAtLocation(usage.symbol, usage.node);
+                // Widen the type so we don't emit nonsense annotations like "function fn(x: 3) {"
+                type = checker.getBaseTypeOfLiteralType(type);
+                typeNode = checker.typeToTypeNode(type, node, NodeBuilderFlags.NoTruncation);
+            }
+
+            const paramDecl = createParameter(
+                /*decorators*/ undefined,
+                /*modifiers*/ undefined,
+                /*dotDotDotToken*/ undefined,
+                /*name*/ name,
+                /*questionToken*/ undefined,
+                typeNode
+            );
+            parameters.push(paramDecl);
+            if (usage.usage === Usage.Write) {
+                (writes || (writes = [])).push(usage);
+            }
+            callArguments.push(createIdentifier(name));
+        });
+
+        // Provide explicit return types for contexutally-typed functions
+        // to avoid problems when there are literal types present
+        if (isExpression(node) && !isJS) {
+            const contextualType = checker.getContextualType(node);
+            returnType = checker.typeToTypeNode(contextualType);
+        }
+
+        const { body, returnValueProperty } = transformFunctionBody(node);
+        let newFunction: MethodDeclaration | FunctionDeclaration;
+
+        if (isClassLike(scope)) {
+            // always create private method in TypeScript files
+            const modifiers: Modifier[] = isJS ? [] : [createToken(SyntaxKind.PrivateKeyword)];
+            if (range.facts & RangeFacts.IsAsyncFunction) {
+                modifiers.push(createToken(SyntaxKind.AsyncKeyword));
+            }
+            if (range.facts & RangeFacts.InStaticRegion) {
+                modifiers.push(createToken(SyntaxKind.StaticKeyword));
+            }
+            newFunction = createMethod(
+                /*decorators*/ undefined,

```

Comment:
>/*decorators*/ undefined [](start = 16, length = 24)

We might eventually want to propagate decorators (e.g. `@readonly` on a parameter.